File: lib/gmail-rate-limit.ts
interface GmailErrorShape {
  code?: number | string;
  message?: string;
  response?: {
    status?: number;
    data?: {
      error?: {
        message?: string;
        errors?: Array<{ reason?: string }>;
      };
    };
  };
}

export interface GmailRetryNotice {
  operation: string;
  attempt: number;
  maxRetries: number;
  delayMs: number;
  reason: string;
}

export interface GmailRequestControllerOptions {
  requestsPerSecond: number;
  maxRetries: number;
  baseDelayMs?: number;
  maxDelayMs?: number;
  onRetry?: (notice: GmailRetryNotice) => void;
  sleep?: (milliseconds: number) => Promise<void>;
  now?: () => number;
  random?: () => number;
}

function defaultSleep(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

function errorDetails(error: unknown): {
  status: number | null;
  reasons: string[];
  message: string;
} {
  const candidate = (error && typeof error === "object" ? error : {}) as GmailErrorShape;
  const numericCode = Number(candidate.code);
  const status = candidate.response?.status ??
    (Number.isFinite(numericCode) ? numericCode : null);
  const reasons = (candidate.response?.data?.error?.errors || [])
    .map((entry) => entry.reason || "")
    .filter(Boolean);
  const message =
    candidate.response?.data?.error?.message ||
    candidate.message ||
    String(error);
  return { status, reasons, message };
}

export function isRetryableGmailError(error: unknown): boolean {
  const details = errorDetails(error);
  if (details.status === 429) return true;
  if (details.status != null && details.status >= 500 && details.status <= 504) {
    return true;
  }
  if (details.status !== 403) return false;

  const retryableReasons = new Set([
    "rateLimitExceeded",
    "userRateLimitExceeded",
    "backendError",
  ]);
  if (details.reasons.some((reason) => retryableReasons.has(reason))) return true;
  return /quota exceeded|rate limit|user-rate limit/i.test(details.message);
}

export class GmailRequestController {
  private readonly intervalMs: number;
  private readonly maxRetries: number;
  private readonly baseDelayMs: number;
  private readonly maxDelayMs: number;
  private readonly onRetry: ((notice: GmailRetryNotice) => void) | undefined;
  private readonly sleep: (milliseconds: number) => Promise<void>;
  private readonly now: () => number;
  private readonly random: () => number;
  private nextStartAt = 0;
  private cooldownUntil = 0;
  private reservation: Promise<void> = Promise.resolve();

  constructor(options: GmailRequestControllerOptions) {
    if (!Number.isFinite(options.requestsPerSecond) || options.requestsPerSecond <= 0) {
      throw new Error("requestsPerSecond must be greater than zero");
    }
    this.intervalMs = Math.ceil(1_000 / options.requestsPerSecond);
    this.maxRetries = options.maxRetries;
    this.baseDelayMs = options.baseDelayMs ?? 1_000;
    this.maxDelayMs = options.maxDelayMs ?? 60_000;
    this.onRetry = options.onRetry;
    this.sleep = options.sleep ?? defaultSleep;
    this.now = options.now ?? Date.now;
    this.random = options.random ?? Math.random;
  }

  private async waitForTurn(): Promise<void> {
    const turn = this.reservation.then(async () => {
      while (true) {
        const waitUntil = Math.max(this.nextStartAt, this.cooldownUntil);
        const waitMs = Math.max(0, waitUntil - this.now());
        if (waitMs <= 0) break;
        await this.sleep(waitMs);
      }
      this.nextStartAt = this.now() + this.intervalMs;
    });
    this.reservation = turn.catch(() => undefined);
    await turn;
  }

  async run<T>(operation: string, request: () => Promise<T>): Promise<T> {
    for (let attempt = 0; ; attempt += 1) {
      await this.waitForTurn();
      try {
        return await request();
      } catch (error) {
        if (attempt >= this.maxRetries || !isRetryableGmailError(error)) {
          throw error;
        }
        const jitterMs = Math.floor(this.random() * 1_000);
        const delayMs = Math.min(
          this.maxDelayMs,
          this.baseDelayMs * 2 ** attempt + jitterMs,
        );
        const details = errorDetails(error);
        this.onRetry?.({
          operation,
          attempt: attempt + 1,
          maxRetries: this.maxRetries,
          delayMs,
          reason: details.message,
        });
        this.cooldownUntil = Math.max(this.cooldownUntil, this.now() + delayMs);
      }
    }
  }
}

File: src/email_automation_jev/gmail_rate_limit.py
from __future__ import annotations

import random
import threading
import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import TypeVar

T = TypeVar("T")


@dataclass(frozen=True)
class GmailRetryNotice:
    operation: str
    attempt: int
    max_retries: int
    delay_seconds: float
    reason: str


def _error_details(error: BaseException) -> tuple[int | None, list[str], str]:
    status = getattr(error, "status_code", None) or getattr(error, "status", None)
    response = getattr(error, "resp", None)
    if response is not None:
        status = getattr(response, "status", status)
    content = getattr(error, "content", b"")
    message = str(error)
    reasons: list[str] = []
    if isinstance(content, bytes):
        content = content.decode("utf-8", errors="replace")
    if isinstance(content, str) and content:
        message = content
        for reason in ("rateLimitExceeded", "userRateLimitExceeded", "backendError"):
            if reason in content:
                reasons.append(reason)
    try:
        numeric_status = int(status) if status is not None else None
    except (TypeError, ValueError):
        numeric_status = None
    return numeric_status, reasons, message


def is_retryable_gmail_error(error: BaseException) -> bool:
    status, reasons, message = _error_details(error)
    if status == 429 or (status is not None and 500 <= status <= 504):
        return True
    if status != 403:
        return False
    if set(reasons) & {"rateLimitExceeded", "userRateLimitExceeded", "backendError"}:
        return True
    lowered = message.lower()
    return "quota exceeded" in lowered or "rate limit" in lowered or "user-rate limit" in lowered


class GmailRequestController:
    """Serializes request starts and retries transient Gmail quota failures."""

    def __init__(
        self,
        requests_per_second: float,
        max_retries: int,
        *,
        base_delay_seconds: float = 1,
        max_delay_seconds: float = 60,
        on_retry: Callable[[GmailRetryNotice], None] | None = None,
        sleep: Callable[[float], None] = time.sleep,
        now: Callable[[], float] = time.monotonic,
        random_value: Callable[[], float] = random.random,
    ) -> None:
        if requests_per_second <= 0:
            raise ValueError("requests_per_second must be greater than zero")
        self.interval = 1 / requests_per_second
        self.max_retries = max_retries
        self.base_delay = base_delay_seconds
        self.max_delay = max_delay_seconds
        self.on_retry = on_retry
        self.sleep = sleep
        self.now = now
        self.random_value = random_value
        self.next_start_at = 0.0
        self.cooldown_until = 0.0
        self._lock = threading.Lock()

    def _wait_for_turn(self) -> None:
        with self._lock:
            wait_until = max(self.next_start_at, self.cooldown_until)
            wait_seconds = max(0.0, wait_until - self.now())
            if wait_seconds:
                self.sleep(wait_seconds)
            self.next_start_at = self.now() + self.interval

    def run(self, operation: str, request: Callable[[], T]) -> T:
        attempt = 0
        while True:
            self._wait_for_turn()
            try:
                return request()
            except BaseException as error:
                if attempt >= self.max_retries or not is_retryable_gmail_error(error):
                    raise
                delay = min(
                    self.max_delay,
                    self.base_delay * (2**attempt) + self.random_value(),
                )
                _, _, reason = _error_details(error)
                if self.on_retry:
                    self.on_retry(GmailRetryNotice(operation, attempt + 1, self.max_retries, delay, reason))
                self.cooldown_until = max(self.cooldown_until, self.now() + delay)
                attempt += 1
